design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+To make possible to specify the remediation strategy for different failure cases and for nodes that run on different providers, this proposal introduces **MachineRemediationRequest** controller that should monitor all **MachineRemediationRequest** CR's and apply the remediation strategy according to details specified under the CR.
+When the **MachineRemediationRequest** will apply the remediation strategy, one who is initiating it will be the **MachineHealthCheck** controller. It will monitor nodes, and for each unhealthy node will create **MachineRemediationRequest** CR, if after the specified timeout the **MachineRemediationRequest** will not have the `remediationState` equal to `Succeeded`, the **MachineHealthCheck** controller will notify about the remediation failure, and can try to use different type of the remediation strategy.
+
+#### MachineRemediationRequest spec API
+
+This proposal introduces a new API type: **MachineRemediationRequest**, that makes possible to create simple **MachineRemediationRequest** object with the type of the remediation strategy for the specific machine.
+
+##### Example
+
+```yaml
+apiVersion: machineremediationrequest.metal3.io/v1alpha1
+kind: MachineRemediationRequest
+metadata:
+  name: mrr-test
+  namespace: openshift-machine-api
+spec:
+  type: reboot
+  machineName: test-machine
+```
+
+#### MachineRemediationRequest status API
+
+**MachineRemediationRequest** status will show what the state the remediation operation has and the time when the remediation operation started.
+
+```yaml
+status:
+  remediationState: InProgress
+  remediationStartTime: Thu, 20 Jun 2019 03:38:39 -0400
+```
+
+### Risks and Mitigations
+
+It can introduce some integration complexity between **MachineHealthCheck** and **MachineRemediationRequest** controllers,
+that will make harder to debug and to test things.
+
+## Design Details
+
+### Work Items
+
+* Create new **MachineRemediationRequest** CRD.
+* Write a new controller that will watch **MachineRemediationRequest** CR's.
+* Write functional tests under <https://github.com/openshift/cluster-api-actuator-pkg> repository.
+* Create an operator to deploy the **MachineRemediationRequest** controller.
+* Create integration with the OLM.
+
+### Dependencies
+
+N/A
+
+### Test Plan
+
+As basic tests, it will be enough to create **MachineRemediationRequest** CR's with the reboot type and verify that the machine really rebooted. In the future, we will need to test integration between **MachineHealthCheck** and **MachineRemediationRequest** controllers to verify that our final fencing solution works as expected.
+
+### Upgrade / Downgrade Strategy
+
+Should be managed by OLM.
+
+### Version Skew Strategy
+
+The **MachineRemediationRequest** controller version should be the same as **MachineHealthCheck** version.
+
+## Alternatives
```

Comment:
Another alternative worth capturing and evaluating here: this could be implemented in the machine actuator as part of the machine controller. It's already watching Machine resources of course, and it already knows how to interact with whatever the local management API is that would carry out remediation.

In this case, the reboot request could be reflected as an annotation on the Machine. Status would be a little trickier, but it's not unheard of to reflect status in an annotation. So you could have a `reboot/request` annotation and a `reboot/last` annotation for example, whose values might be timestamps.

Without direct support in the machine API, this would not be as robust or clear of a solution as your proposal. But it would be simple to implement and would not require a new controller. It also seems like a good behavioral fit for the machine controller.

Reply:
This was one of the original implementation ideas, but there wasn't a clear way to tell the actuator whether reboot or deletion was required.  Applying the same signalling idea proposed here to this concept would work, however I suspect mixing the reboot state machine into the actuator will get messy

Reply:
I wouldn't worry about the complexity of the machine controller. There's not a whole lot to it today, and adding reboot logic would not be a burden by any means. People are making far more complex controllers as a normal practice these days.

Reply:
I will add this as an option.